Pod Logs Fullscreen

dtaylor113 opened the pull request:
**Expand:**
![image](https://user-images.githubusercontent.com/12733153/44226148-52eb8f80-a15d-11e8-9dd5-6841fa0d55d3.png)

**Collapse:**
_* screen capture tool is truncating margins, toolbar has a 10px margin/padding all around_
![image](https://user-images.githubusercontent.com/12733153/44226166-5b43ca80-a15d-11e8-8686-2571386b9073.png)

JIRA: https://jira.coreos.com/browse/CONSOLE-610

@openshift/team-ux-review

TheRealJon commented on frontend/public/components/utils/resource-log.jsx:
```diff
@@ -26,8 +27,26 @@ const streamStatusMessages = {
 // Component for log stream controls
-const LogControls = ({dropdown, onDownload, status, toggleStreaming}) => {
+const LogControls = ({dropdown, onDownload, toggleFullscreen, status, toggleStreaming}) => {
+  const expandCompressClass = classNames({
```
Since `fa` will always be used, this can be: 
```js
classNames( 'fa', {
    'fa-expand': !isFullscreen(), 
    'fa-compress': isFullscreen()
});
```

TheRealJon commented on frontend/public/components/utils/resource-log.jsx:
```diff
@@ -69,6 +94,11 @@ export class ResourceLog extends SafetyFirst {
+      document.documentElement.requestFullScreen ||
+      document.documentElement.webkitRequestFullScreen ||
+      document.documentElement.mozRequestFullScreen ||
+      document.documentElement.msRequestFullscreen;
```
`document.documentElement.msRequestFullscreen` may not be necessary since we don't officially support IE

dtaylor113 commented:
Build -> Logs isn't working correctly. Seems to be a race condition in `isFullscreen()` and the `document.` assigning/getting a `fullscreenElement`

TheRealJon commented on frontend/public/components/utils/resource-log.jsx:
```diff
@@ -247,17 +299,20 @@ export class ResourceLog extends SafetyFirst {
-        status={status}
-        updateStatus={this._updateStatus} />
+      <div className="log-console">
+        <LogControls
```
nit: I think I missed this element when I first created this, but I try to keep the props in alphabetical order when they are split across multiple lines like this. This isn't consistent across the repo, but when I encounter them in something I work on, I try to fix it. I don't think it's all that important, but just thought I'd mention it.

dtaylor113 replied:
fixed

TheRealJon commented on frontend/public/components/utils/resource-log.jsx:
```diff
@@ -114,6 +144,28 @@ export class ResourceLog extends SafetyFirst {
+      } else if(document.webkitExitFullscreen) {
+        document.webkitExitFullscreen();
+      } else if (document.msExitFullscreen) {
+        document.msExitFullscreen();
```
`document.msExitFullscreen();` may not be necessary since we don't officially support IE.

benjaminapetersen replied:
Prob if we drop that it would be good to ensure it doesn't error/fall over.  No need to deliberately break. 😄

TheRealJon commented:
One general comment. Does it make sense to track the fullscreen status as state? Also, since `requestFullScreen` is async, but returning a promise is not supported by any browser yet, should we implement something that handles the async nature of this feature? In other words, our own `requestFullscreen` method that will return a promise?

robszumski commented:
Jenkins push

benjaminapetersen commented on frontend/public/components/utils/resource-log.jsx:
```diff
@@ -26,8 +27,26 @@ const streamStatusMessages = {
   [STREAM_ACTIVE]: 'Log streaming...'
 };
 
+const isFullscreen = function() {
```
You may want to consider just collecting all of the `fullScreen` bits up into one object adapter, just to keep the related bits associated:

```JavaScript 
const fullScreenAdapter = {
  isSupported: () => {
    return document.fullscreenElement || 
      document.webkitFullscreenElement || 
      document.mozFullScreenElement;
  },
  enter: (element) => {
    const elem = element.length ? element[0] : element;
    return elem && (document.documentElement.requestFullScreen ||
    document.documentElement.webkitRequestFullScreen ||
    document.documentElement.mozRequestFullScreen ||
    document.documentElement.msRequestFullscreen || 
    function() { }).call(elem);
  },
  exit: () => {
    return (document.exitFullscreen ||
    document.mozCancelFullScreen ||
    document.webkitExitFullscreen || 
    document.msExitFullscreen || 
    function() { })();
  }
}
```
NOTE: I didn't test this extensively, I know there are some quirks around how you call these (context).

benjaminapetersen replied:
Also, not sure if you want the null object pattern with the trailing anonymous functions (in my example), but might be handy.

benjaminapetersen replied:
Leaning towards thinking this could be a good utility to extract into `/utils`, to avoid duplication if any other component needs fullscreen.

dtaylor113 replied:
Hi @TheRealJon, yes, I did have it that way but it wasn't working correctly.  I plan on refactoring the method.

benjaminapetersen commented on frontend/public/components/utils/resource-log.jsx:
```diff
@@ -114,6 +144,28 @@ export class ResourceLog extends SafetyFirst {
 
+  // Toggle currently displayed log content to/from fullscreen
+  _toggleFullscreen () {
+    const logConsole = document.querySelector(".log-console");
```
Is using `document.querySelector(".log-console")` best practice?  I would think there is a better method for this.   Perhaps a [Ref](https://reactjs.org/docs/refs-and-the-dom.html).

benjaminapetersen replied:
Much less code using that util, still curious about this one.

dtaylor113 replied:
hi, yes, looking into using Refs, thanks

dtaylor113 replied:
Hi, fixed to use Refs

dtaylor113 commented:
> ...our own requestFullscreen method that will return a promise?
Hi @TheRealJon, ah, this must be why it doesn't work all the time, I forgot about the async aspect, thanks, I'll look at refactoring.

spadgett commented:
/hold
for feature freeze

dtaylor113 commented:
Hi @TheRealJon, @benjaminapetersen,   I found a good util which seems to do all the fullscreen variations and promises: https://github.com/shisama/toggle-fullscreen.   Updated the code, seems to be much cleaner.

spadgett commented on frontend/package.json:
```diff
@@ -87,6 +87,7 @@
     "sanitize-html": "1.x",
     "showdown": "1.8.x",
     "text-encoding": "0.x",
+    "toggle-fullscreen": "^0.3.3",
```
This library isn't widely used or very active (0 stars), which makes me a little hesitant to add it.

I'd suggest something like https://github.com/sindresorhus/screenfull.js

benjaminapetersen replied:
(noticed that as well and was a little concerned)

dtaylor113 replied:
Hi, switched to screenfull, thanks

dtaylor113 commented:
Switched to suggested lib/util.  I believe I addressed all comments -thanks

tlwu2013 commented:
I personally feel the double buttons ("Download"/"Expand") are visually too heavy. I'll vote for replacing them with button link again. 

**Mocked screenshot**:
![pod-details__logs-tab](https://user-images.githubusercontent.com/5903705/42771635-b4f95638-88dc-11e8-9a9f-93d7f9cc7336.png)


**OpenShift 3.10** (for reference):
<img width="1391" alt="app-console__log-tab" src="https://user-images.githubusercontent.com/5903705/42771376-dd07d5ec-88db-11e8-8d6b-0349afbdbada.png">

Thoughts? @openshift/team-ux-review  Thanks!

spadgett commented on frontend/public/components/utils/resource-log.jsx:
```diff
@@ -4,9 +4,11 @@ import { Base64 } from 'js-base64';
 import { modelFor, resourceURL } from '../../module/k8s';
 import { SafetyFirst } from '../safety-first';
 import { WSFactory } from '../../module/ws-factory';
+import * as screenfull from 'screenfull';
```
👍

dtaylor113 replied:
Hi @spadgett, fixed these issues.

dtaylor113 commented:
@tlwu2013, replaced buttons with links, pls see updated screenshots in this PR's description -thanks

TheRealJon commented on frontend/public/components/utils/_log-window.scss:
```diff
@@ -4,8 +4,21 @@ $color-log-window-text: $color-pf-black-150;
+.log-console {
+  height: 100%;
+  width: 100%;
+  &.fullscreen {
```
should it be `log-console--fullscreen` since this is more like a modifier than a component of the the log-console?

spadgett replied:
Yes I got it backwards :(

spadgett commented on frontend/public/components/utils/resource-log.jsx:
```diff
@@ -247,17 +283,21 @@ export class ResourceLog extends SafetyFirst {
-        bufferFull={bufferFull}
-        status={status}
-        updateStatus={this._updateStatus} />
+      <div ref={this._logConsole} className={classNames('log-console', {'log-console--toolbar__fullscreen': isFullscreen})}>
```
Per BEM, this class would need to go on the toolbar element itself

spadgett replied:
Also per @TheRealJon it should be `log-console__toolbar--fullscreen`. (I got the BEM naming wrong.)

spadgett commented on frontend/public/components/utils/_log-window.scss:
```diff
@@ -4,8 +4,21 @@ $color-log-window-text: $color-pf-black-150;
+.log-console {
+  height: 100%;
+  width: 100%;
+  &.log-console--toolbar__fullscreen {
```
Thanks, this is better than my suggestion

TheRealJon commented on frontend/public/components/utils/resource-log.jsx:
```diff
@@ -53,21 +64,24 @@ export class ResourceLog extends SafetyFirst {
     this._restartStream = this._restartStream.bind(this);
     this._toggleStreaming = this._toggleStreaming.bind(this);
     this._updateStatus = this._updateStatus.bind(this);
+    this._logConsole = React.createRef();
```
Could do `this._resourceLogRef = React.createRef();` or something along those lines?

dtaylor113 commented:
Hi @tlwu2013, added the divider:
![image](https://user-images.githubusercontent.com/12733153/42822248-91442764-89a8-11e8-82fe-71cec78e5843.png)

@TheRealJon, @spadgett, updated the scss, hopefully I got the BEM correct. -thanks

spadgett commented on frontend/public/components/utils/_resource-log.scss:
```diff
@@ -0,0 +1,13 @@
+.resource-log {
+  height: 100%;
+  width: 100%;
+  &.log-console--fullscreen {
```
I don't think this needs to be nested under .resource-log

TheRealJon replied:
Shouldn't this also be `&.resource-log--fullscreen?

dtaylor113 replied:
Hi, fixed, still learning BEM, thanks reviewing

TheRealJon replied:
... or if not nested, `.resource-log.resource-log--fullscreen`

spadgett commented on frontend/public/components/utils/_resource-log.scss:
```diff
@@ -0,0 +1,13 @@
+    }
+  }
+  .pairs-list__action-divider {
+    padding: 0;
```
This is not the pairs-list block. We either need to generalize this class or have a special class for the log viewer

dtaylor113 replied:
fixed, third time's a charm :-)

TheRealJon replied:
You could probably make this a part of the `co-toolbar` styles in `frontend/styles/_common.scss`. Something like `.co-toolbar__item--divider`.

TheRealJon commented on frontend/public/components/utils/_resource-log.scss:
```diff
@@ -0,0 +1,16 @@
+  }
+}
+
+.resource-log__action-divider {
```
I think this should be part of the `.co-toolbar` styles [here](https://github.com/openshift/console/blob/20b586c34ef0bfad7fe2f8bef5243f10eb650d43/frontend/public/style/_common.scss#L227-L268)

so this would become `.co-toolbar__item--divider`

dtaylor113 commented:
Hi, I had the HTML wrong, I included 'Download | Expand' in a single `.co-toolbar__item`. I just updated the HTML to include Download, the divider, and Expand each in their own `co-toolbar__item` div, but that adds a 15px left padding to each right-hand item in the toolbar:
![image](https://user-images.githubusercontent.com/12733153/42827719-17ed9e6a-89b5-11e8-958e-12be0bc2ade3.png)

This matches the 15px right padding for the controls on the left side of the toolbar (between 'Log streaming...' and the dropdown).

@tlwu2013, should the right-hand side button links/dividers be closer together or do we want 15px space between them?

**Before:**
![image](https://user-images.githubusercontent.com/12733153/42829801-30417600-89b8-11e8-8fbb-9b47dd94769b.png)

**After corrected HTML:**
![image](https://user-images.githubusercontent.com/12733153/42829557-d32de124-89b7-11e8-91e7-29cce75c5efe.png)


Thanks

dtaylor113 commented:
Pushed code changes/fixes.

spadgett commented on frontend/public/style/_common.scss:
```diff
@@ -231,6 +231,11 @@ input[type=number]::-webkit-outer-spin-button {
 }
 
+.co-toolbar__item--divider {
+  color: $color-pf-black-500;
```
We should have some way of ensuring that this and `pairs-list__action-divider` use the same color (and maybe other styles). So either a common var for color or perhaps a general `co-action-divider` class that is used in both places.

dtaylor113 replied:
Now using a general `co-action-divider` class in both places.

spadgett commented on frontend/public/style/_common.scss:
```diff
@@ -202,6 +202,18 @@ input[type=number]::-webkit-outer-spin-button {
 }
 
+.co-action-divider {
+  display: none;
```
Let's not assume that links will stack and the divider should be hidden in all cases. That really depends on context.

dtaylor113 replied:
So it sounds like we shouldn't try to have a common `co-action-divider` class then?  It sounds like I should just use a common color variable in both `pairs-list__action-divider` and `.co-toolbar__item--divider`. rather than trying to use `co-action-divider` in both places.

spadgett replied:
I still think it's the right thing to do.

I'd leave the common rule and keep the pairs rule with the media queries.

spadgett replied:
If we don't have a common rule for the dividers, the styles will drift from page to page.

dtaylor113 replied:
> I'd leave the common rule and keep the pairs rule with the media queries.

Ok, will do, but should the resource log hide the divider at mobile?

dtaylor113 replied:
Ok, `co-action-divider` just has the color, `pairs-list__action-divider` hides the divider at mobile.  I can add `.co-toolbar__item--divider` to also hide the divider at mobile if that is what we decide.

spadgett commented on frontend/public/style/_common.scss:
```diff
@@ -202,6 +202,18 @@ input[type=number]::-webkit-outer-spin-button {
+    display: inline;
+    color: $color-pf-black-600;
+    padding-right: 10px;
+    &.no-padding {
```
This is because btn-link has padding?

I'd rather have a class that removes padding for btn-link than a special case on the divider.

dtaylor113 replied:
Yes, `pairs-list` uses `<span>`'s whereas `resource-log` uses `<buttons>`

spadgett replied:
How hard is it to switch pairs list to btn-link? Then we can remove the padding styles entirely and it fixes an accessibility issue.

@jcaianirh FYI

dtaylor113 replied:
> How hard is it to switch pairs list to btn-link? Then we can remove the padding styles entirely and it fixes an accessibility issue.

Switched pairs list to use `<button>`'s, consolidated scss classes.

spadgett commented on frontend/public/style/_common.scss:
```diff
@@ -202,6 +202,17 @@ input[type=number]::-webkit-outer-spin-button {
+  color: $color-pf-black-600;
+}
+
+.co-action-divider--hide-at-mobile {
```
Any reason not to just use `hidden-xs`?

dtaylor113 replied:
Nope :-)

dtaylor113 commented:
/retest

spadgett commented:
jenkins rebuild

spadgett commented:
@dtaylor113 if you rebase, i believe the integration tests will start passing again

dtaylor113 commented:
jenkins rebuild

dtaylor113 commented:
Hi @spadgett, still failing :-(

GUI Tests:  
```
Failed Specs:
1) Kubernetes resource CRUD operations Namespace : deletes the namespace
   WebDriverError: unknown error: Element <a>...</a> is not clickable at point (345, 1187). Other element would receive the click: <html lang="en" class="no-js">...</html>" & 

BEGIN BROWSER LOGS
   (most lines end in) Failed to load resource: the server responded with a status of 404 (Not Found)
```
Jenkins: 
```
org.acegisecurity.AccessDeniedException: Please login to access job console-build
```

spadgett commented:
Different flake than we fixed :/

spadgett commented:
jenkins rebuild

dtaylor113 commented:
Hi @spadgett, anything I should do to move the PR along?  -thanks

benjaminapetersen commented:
@dtaylor113 lgtm at this point, though im not sure if we are merging new features yet...

spadgett commented:
> anything I should do to move the PR along? -thanks

We're in feature freeze, so this will need to wait until we branch for new feature work.

spadgett commented:
jenkins rebuild

spadgett commented:
/hold cancel

spadgett commented:
@dtaylor113 The logs don't extend to the bottom of the page in fullscreen for me:

![screen shot 2018-08-15 at 4 05 28 pm](https://user-images.githubusercontent.com/1167259/44170359-1b211100-a0a5-11e8-93f6-4587867549b7.png)

beanh66 commented:
@dtaylor113 Looking good! Minor thing, can we change that "Log Streaming..." and "Log Stream Paused" text to say "Container:" instead? We just made this change to the Environment page as well and would like to be consistent on the Logs and Terminal tabs.

dtaylor113 commented:
> The logs don't extend to the bottom of the page in fullscreen for me
Hi @spadgett, that appears to be a screenshot of the Pod Terminal, not logs :-)

dtaylor113 commented:
I just ran this locally in Chrome and I'm seeing:
![image](https://user-images.githubusercontent.com/12733153/44216302-9dabde00-a142-11e8-915f-7a3a843e7709.png)

spadgett commented on frontend/public/components/utils/resource-log.jsx:
```diff
@@ -23,12 +25,16 @@ export const LOG_SOURCE_WAITING = 'waiting';
-  [STREAM_PAUSED]: 'Log stream paused.',
-  [STREAM_ACTIVE]: 'Log streaming...'
+  [STREAM_PAUSED]: 'Container:',
+  [STREAM_ACTIVE]: 'Container:'
```
If we're going to make this change, we should change all of these messages and get rid of this object. I'd prefer to look at this separately if you don't mind @beanh66

dtaylor113 replied:
It depends on if we still want to dynamically show there
```
[STREAM_EOF]: 'Log stream ended.',
[STREAM_LOADING]: 'Loading log...',
```

spadgett replied:
If we want "Container:" we should always show that, but then we need to find another place for these messages.

I'd rather address in a follow on since it's an unrelated change and this PR is almost ready.

dtaylor113 replied:
Ok, reverted.

spadgett commented:
I opened https://jira.coreos.com/browse/CONSOLE-713 to address the dropdown labels

spadgett commented:
jenkins rebuild

spadgett commented:
jenkins rebuild

spadgett commented:
Last failure was flake: https://jira.coreos.com/browse/CONSOLE-697

spadgett commented:
jenkins rebuild

spadgett commented:
jenkins rebuild

spadgett commented:
/lgtm

spadgett commented:
jenkins rebuild

spadgett commented:
@dtaylor113 The test failures were legitimate. I pushed a fix to your branch.

/lgtm
